polkadot/node/collation-generation/src/lib.rs

```diff
@@ -420,90 +372,89 @@ async fn handle_new_activations<Context>(
 							validation_code_hash,
 							n_validators,
 							core_index,
+							session_index,
 						},
 						task_config.key.clone(),
 						&mut task_sender,
 						result_sender,
 						&metrics,
+						session_info.v2_receipts,
+						&transposed_claim_queue,
 					)
-					.await;
+					.await
+					{
+						gum::error!(
+							target: LOG_TARGET,
+							"Failed to construct and distribute collation: {}",
+							err
+						);
+						return
+					}
 
 					// Chain the collations. All else stays the same as we build the chained
 					// collation on same relay parent.
 					validation_data.parent_head = parent_head;
 				}
 			}),
 		)?;
-	}
 
-	Ok(())
+		Ok(())
+	}
 }
 
-#[overseer::contextbounds(CollationGeneration, prefix = self::overseer)]
-async fn handle_submit_collation<Context>(
-	params: SubmitCollationParams,
-	config: &CollationGenerationConfig,
-	ctx: &mut Context,
-	metrics: &Metrics,
-) -> crate::error::Result<()> {
-	let _timer = metrics.time_submit_collation();
+#[overseer::subsystem(CollationGeneration, error=SubsystemError, prefix=self::overseer)]
+impl<Context> CollationGenerationSubsystem {
+	fn start(self, ctx: Context) -> SpawnedSubsystem {
+		let future = async move {
+			self.run(ctx).await;
+			Ok(())
+		}
+		.boxed();
 
-	let SubmitCollationParams {
-		relay_parent,
-		collation,
-		parent_head,
-		validation_code_hash,
-		result_sender,
-		core_index,
-	} = params;
+		SpawnedSubsystem { name: "collation-generation-subsystem", future }
+	}
+}
 
-	let validators = request_validators(relay_parent, ctx.sender()).await.await??;
-	let n_validators = validators.len();
+#[derive(Clone)]
+struct SessionInfo {
```

Comment:
I'd pick a different name for this struct. Before seeing the declaratioin I thought `SessionInfo` is [the one from primitives](https://github.com/paritytech/polkadot-sdk/blob/4c618a83d33281fe96f0e2b68a111ed227af22c0/polkadot/primitives/src/v8/mod.rs#L1886)

Reply:
Good point. Renamed to `PerSessionInfo`